utils/polus-bfio-util/bfio/bfio.py

```diff
@@ -55,194 +82,356 @@ class BioReader():
                                     x_range = x_max - x
-                                    for y in range(Y[0],Y[1],self._TILE_SIZE):
-                                        y_max = np.min([y+self._TILE_SIZE,Y[1]])
+                                    for y in range(Y[0], Y[1], self._TILE_SIZE):
+                                        y_max = np.min(
+                                            [y+self._TILE_SIZE, Y[1]])
                                         y_range = y_max - y
-                                        I[y-Y[0]:y_max-Y[0],x-X[0]:x_max-X[0],zi,ci,ti] = reader.read(c=c,z=z,t=t,rescale=False,XYWH=(x,y,x_range,y_range))
-            
+                                        I[y-Y[0]:y_max-Y[0], x-X[0]:x_max-X[0], zi, ci, ti] = reader.read(
+                                            c=c, z=z, t=t, rescale=False, XYWH=(x, y, x_range, y_range))
+
         return I
-    
+
+
 class BioWriter():
-    # Note: the javabridge connection must be started before initializing a 
-    # BioReader object. The reason for this is that the java VM can only be
-    # initialized once. Doing it within the code would cause it to run possibly
-    # without shutting down, so the javabridge connection must be handled
-    # outside of the class.
+    """BioWriter Write OME tiled tiffs using Bioformats
+
+    This class handles writing data to OME tiled tiff format using the 
+    OME Bioformats tool. Like the BioReader class, it handles writing
+    large images (>2GB).
+    
+    Once the class is initialized, and once the write_image() function
+    is called at least once, most of the methods to set the metadata
+    information will throw an error. Therefore, make sure to set all
+    metadata before image writing is started.
+    
+    Note: The javabridge is not handled by the BioReader class. It must be
+          initialized prior to using the BioReader class, and must be closed
+          before the program terminates. An example is provided in read_image().
+    
+    For for information, visit the Bioformats page:
+    https://www.openmicroscopy.org/bio-formats/
+
+    Methods:
+        BioReader(file_path,image,X,Y,Z,C,T,metadata): See __init__ for details
+        pixel_type(dtype): Gets/sets the pixel type (e.g. uint8)
+        channel_names(cnames): Gets/sets the names of each channel
+        num_x(X): Get/set number of pixels in the x-dimension (width)
+        num_y(Y): Get/set number of pixels in the y-dimension (height)
+        num_z(Z): Get/set number of pixels in the z-dimension (depth)
+        num_c(C): Get/set number of channels in the image
+        num_t(T): Get/set number of timepoints in the image
+        physical_size_x(psize,units): Get/set the physical size of the x-dimension
+        physical_size_y(psize,units): Get/set the physical size of the y-dimension
+        physical_size_z(psize,units): Get/set the physical size of the z-dimension
+        write_image(image,X,Y,Z,C,T): Writes the 5d image array to file
```

Comment:
Add series

Reply:
`series` is not a method. This is a list of methods.